### PR TITLE
✨#459: connect invite UI to API

### PR DESCRIPTION
**Pull request by samrichca**
### Consent UI
- move API calls to services folder.
- add form submission to invite form.
- add a form wrapper and suspense boundary to the invite form.
- add a `qaLogger` function
- transform `participantPreferredName` to `undefined` if it's empty

### Consent API
- change consentToBeContacted to `literal(true)`, remove `optional`, in ClinicianInviteResponse

### Types
- update vitest to continue using the vscode plugin

## PR Readiness Checklist

- [x] "Expected Outcome(s)" in ticket have been met
- [x] Ticket number included in PR title
- [x] Connected ticket to PR
- [x] Labels added to PR for service name (`consent-api`, `data-mapper`, etc...), type (`chore`, `documentation`, etc...), status (`draft`, `on-hold`, etc...) **if applicable**
- [x] Manual testing completed
- [x] Builds locally without errors or warnings
- [x] Tests are updated (if required) and passing
- [ ] PR feedback has been addressed **if applicable**
- [x] New environment variables added to `.env.schema` files, `README.md`
- [x] Added copyrights to new files

### CODE REVIEW

File: apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx [hunk 1]
```diff
@@ -86,6 +89,18 @@ const ClinicianInviteFormComponent = ({
 		mode: 'onBlur',
 		resolver: zodResolver(ClinicianInviteRequest),
 		shouldUnregister: true,
+		defaultValues: {
```
Comment (samrichca) [inline apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx:92]: temp for demo

File: packages/types/src/services/consentApi/responses/ClinicianInvite.ts [hunk 2]
```diff
@@ -25,8 +25,8 @@ import { InviteEntity, ClinicianInviteBase } from '../../../entities/index.js';
 
 export const ClinicianInviteResponse = InviteEntity.merge(ClinicianInviteBase).merge(
 	z.object({
+		consentToBeContacted: z.literal(true),
```
Comment (samrichca) [inline packages/types/src/services/consentApi/responses/ClinicianInvite.ts:28]: this is required in the form so I think it shouldn't be optional in the response

File: apps/consent-ui/src/components/views/Invite/ClinicianInviteFormWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+import { CONSENT_GROUPS } from 'types/entities';
+
+import { ValidLanguage, getTranslation } from 'src/i18n';
+
+import ClinicianInviteFormComponent from './ClinicianInviteForm';
+import { ConsentGroupOption } from './ClinicianInviteForm/types';
+
+const ClinicianInviteFormWrapper = async ({ currentLang }: { currentLang: ValidLanguage }) => {
```
Comment (samrichca) [inline apps/consent-ui/src/components/views/Invite/ClinicianInviteFormWrapper.tsx:27]: I had to make a wrapper for the register form and will need it for consent steps - to fetch data for that specific form - so I did it here too for consistency.

File: apps/consent-ui/src/services/api/requests/postInvite.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/*
```
Comment (samrichca) [inline apps/consent-ui/src/services/api/requests/postInvite.ts:1]: gonna start putting API requests in this folder. the register "getInvite" request should be in here as well but not on this PR

File: apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx [hunk 5]
```diff
@@ -110,17 +125,22 @@ const ClinicianInviteFormComponent = ({
...
-			// on success, go to homepage & show success message
-			showNotification({ page: 'home', notification: 'inviteSent' });
-			router.push(getLocalizedRoute(currentLang, 'home'));
+			startTransition(async () => {
+				const inviteResult = await handleSubmitInvite(data, recaptchaToken);
+				if (inviteResult.status === 'SUCCESS') {
+					showNotification({ page: 'home', notification: 'inviteSent' });
+					router.push(getLocalizedRoute(currentLang, 'home'));
+				} else {
+					// TODO: Adding a log for testing purpose, but should implement proper error handling
```
Comment (samrichca) [inline apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx:140]: gonna ask about this at f2f

File: apps/consent-ui/src/services/api/requests/postInvite.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import { API } from 'src/constants';
+import { RecaptchaToken } from 'src/hooks/useRecaptcha';
+
+import consentApiFetch from '../axios/consentApiFetch';
+import catchAxiosError from '../axios/catchAxiosError';
+
+/**
+ * Send invite form data to consent API.
+ * @param data ClinicianInviteRequest
+ * @returns ClinicianInviteResponse | null
```
Comment (samrichca) [inline apps/consent-ui/src/services/api/requests/postInvite.ts:31]: have to update this line

File: packages/types/src/services/consentApi/responses/ClinicianInvite.ts (hunk 2 above)
Comment (anncatton) [reply packages/types/src/services/consentApi/responses/ClinicianInvite.ts:28]: Agree with this. This also made me realize we are handling the response further down incorrectly - it's being treated as nullable in the DB, which it is not - it's a non-nullable boolean field. We are just enforcing that the request value *must be true* for any service to accept it. That means we should be able to modify the Data Mapper `ConsentClinicianInvite` and the Consent DAS `ConsentClinicianInvite` response types to have this field as `literal(true)` without the conversion from `nullable => undefined`. You can update in this PR.

File: packages/types/src/services/consentUi/requests/ClinicianInvite.ts [hunk 7]
```diff
@@ -84,3 +86,6 @@ export const ClinicianInviteRequest = InviteClinicianFields.and(ConsentToBeConta
 		}),
 	);
 export type ClinicianInviteRequest = z.infer<typeof ClinicianInviteRequest>;
+
+export type PostInviteResult = Promise<{ data?: ClinicianInviteResponse; error?: unknown }>;
```
Comment (anncatton) [inline packages/types/src/services/consentUi/requests/ClinicianInvite.ts:90]: See my comment above about reworking/renaming this type.

File: apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx [hunk 8]
```diff
@@ -110,17 +125,23 @@ const ClinicianInviteFormComponent = ({
...
-
-			// on success, go to homepage & show success message
-			showNotification({ page: 'home', notification: 'inviteSent' });
-			router.push(getLocalizedRoute(currentLang, 'home'));
+			startTransition(async () => {
+				const inviteResult = await handleSubmitInvite(data, recaptchaToken);
+				if (inviteResult.status === 'SUCCESS') {
+					showNotification({ page: 'home', notification: 'inviteSent' });
+					router.push(getLocalizedRoute(currentLang, 'home'));
+					console.info('QA - INVITE CREATED. ID:', inviteResult.data.id);
```
Comment (anncatton) [inline apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx:139]: I think we can use the logger package to help us call out these QA logs. Just need to set up a top level logger like we have in the other services (i.e. [consent-api](https://github.com/OHCRN/platform/blob/develop/apps/consent-api/src/logger.ts), with `forModule(Consent UI) and then we can have a named submodule for logs like these, as `const qaLogger = logger.forModule('QA')`. Should then be easier to trace and remove when we have other visible UI errors for testing.

**Review by anncatton — COMMENTED** [review]
Nice to see the api hookup working! I think we can do some cleanup here tho that should help us with subsequent UI -> API work.

File: apps/consent-ui/src/components/utils.ts [hunk 9]
```diff
@@ -80,3 +80,17 @@ export const getUserType = (consentGroup: ConsentGroup): UserType | undefined =>
 
 	return userType;
 };
+
+/**
+ * Convert empty strings to undefined for optional form fields, e.g. for API request body data.
+ * Can have the same input and return type because the affected fields should allow undefined values.
+ * @param formData data object from the form
+ */
+export const convertEmptyStringsToUndefined = <T extends object>(data: T): T => {
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/utils.ts:89]: good idea. I discovered it's changing empty or whitespace fields to undefined on submit, without the transform or the convert... function

image

[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/utils.ts]: I tried this but it didn't change the data.

File: packages/types/src/services/consentUi/responses/ClinicianInvite.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { SystemError } from '../../../httpResponses/index.js';
+import { ClinicianInviteResponse } from '../../consentApi/index.js';
+
+export type ClinicianInviteResult = Promise<{ data?: ClinicianInviteResponse; error?: unknown }>;
```
Comment (anncatton) [inline packages/types/src/services/consentUi/responses/ClinicianInvite.ts:23]: i would define the `Response` type only here, and refer to the result type separately in the `handleCreateInvite`. I think it's easier to read that way. I would also refrain from importing a type in one service into another service within the types package. Instead, you can import the `consentApi` `ClinicianInviteResponse` directly into your handler file, and then add in the `Result`/`Failure` code there:
```
// handleCreateInvite.ts
type ClinicianInviteFailureStatus = 'INVITE_EXISTS' | SystemError;
...handler code
```
```
// createInvite.ts
const createInvite =  async (
	data: ClinicianInviteRequest,
	recaptchaToken: RecaptchaToken,
): Promise<{ data?: ClinicianInviteResponse; error?: unknown }>
```

Or even nicer, we could add a generic response type that will integrate the `catchAxiosError` + the applicable data type, like:
```
type HandledResponse<T> = Promise<{ data?: T; error?: unknown }>;
```
and use like:
```
createInvite = async(...):HandledResponse<ClinicianInviteResponse>;
```

**Review by anncatton — COMMENTED** [review]
this is great Sam! I just added one suggestion for making the axios response handling a bit tidier

**Review by anncatton — APPROVED** [review]
🚀

**Pull request merged**